### PR TITLE
Alerting: Add support for running HA using Redis

### PR DESCRIPTION
**Pull request by JohnnyQQQQ**
**What is this feature?**

This feature adds the possibility to use Redis instead of Memberlist for HA Alerting.

Testing this might be difficult, as it involves at least two Grafana instances. I created a straight forward docker compose template that can be used. https://github.com/JohnnyQQQQ/grafana-ha-testing-template

### CODE REVIEW

File: conf/defaults.ini [hunk 1]
```diff
@@ -939,6 +939,26 @@ admin_config_poll_interval = 60s
 # The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
 alertmanager_config_poll_interval = 60s
 
+# The redis server address that should be connected to.
+ha_redis_address =
```
Comment (alexweav) [inline conf/defaults.ini:943]: I think we usually also copy new config keys into `sample.ini` and comment it out.

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go [hunk 2]
```diff
@@ -72,14 +72,40 @@ func NewMultiOrgAlertmanager(cfg *setting.Cfg, configStore AlertingStore, orgSto
...
+			password: cfg.UnifiedAlerting.HARedisPassword,
+			username: cfg.UnifiedAlerting.HARedisUsername,
+			db:       cfg.UnifiedAlerting.HARedisDB,
+		}, clusterLogger, moa.metrics.Registerer, cfg.UnifiedAlerting.HAPushPullInterval)
+		if err != nil {
+			return fmt.Errorf("unable to initialize redis: %w", err)
+		}
+		var ctx context.Context
+		ctx, moa.settleCancel = context.WithTimeout(context.Background(), 30*time.Second)
+		go redisPeer.Settle(ctx, cluster.DefaultGossipInterval*10)
```
Comment (alexweav) [inline pkg/services/ngalert/notifier/multiorg_alertmanager.go:100]: Why `*10`?

We could add a comment explaining, or a const for the 10 so it's easier to understand?

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go [hunk 3]
```diff
@@ -72,14 +72,40 @@ func NewMultiOrgAlertmanager(cfg *setting.Cfg, configStore AlertingStore, orgSto
...
+			prefix:   cfg.UnifiedAlerting.HARedisPrefix,
+			password: cfg.UnifiedAlerting.HARedisPassword,
+			username: cfg.UnifiedAlerting.HARedisUsername,
+			db:       cfg.UnifiedAlerting.HARedisDB,
+		}, clusterLogger, moa.metrics.Registerer, cfg.UnifiedAlerting.HAPushPullInterval)
+		if err != nil {
+			return fmt.Errorf("unable to initialize redis: %w", err)
+		}
+		var ctx context.Context
+		ctx, moa.settleCancel = context.WithTimeout(context.Background(), 30*time.Second)
```
Comment (alexweav) [inline pkg/services/ngalert/notifier/multiorg_alertmanager.go:99]: can we make the 30s timeout a const?

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go [hunk 4]
```diff
@@ -72,14 +72,40 @@ func NewMultiOrgAlertmanager(cfg *setting.Cfg, configStore AlertingStore, orgSto
...
+		go redisPeer.Settle(ctx, cluster.DefaultGossipInterval*10)
+		moa.peer = redisPeer
+		return nil
+	}
+	// Memberlist setup.
 	if len(cfg.UnifiedAlerting.HAPeers) > 0 {
```
Comment (alexweav) [inline pkg/services/ngalert/notifier/multiorg_alertmanager.go:105]: What happens if the user configures _both_ memberlist peers and redis?

**Review by alexweav — COMMENTED** [review]
Mostly focused on setup, only briefly reviewed the peer for now. Just a couple nitpicks so far, dropping them now as this area is a bit separate.

**Review by stevesg — COMMENTED** [review]
Nice work overall! Just a couple of important changes and some suggestions/nits.

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 5]
```diff
@@ -0,0 +1,590 @@
...
+)
+
+type redisPeer struct {
+	name   string
+	redis  *redis.Client
+	prefix string
+	logger log.Logger
+	states map[string]cluster.State
+	subs   map[string]*redis.PubSub
+	mtx    sync.RWMutex
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:51]: Changed 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+
+func (p *redisPeer) Position() int {
+	startTime := time.Now()
+	// When the members array is empty and we fetched our position successfully
+	// prior, we can assume that currently the network is down.
+	// To prevent a spike of duplicate messages, we return for the duration of
+	// lastPositionValidFor the last known position and only failover to position
+	// 0 afterwards if we do not eventually recover.
+	if len(p.Members()) == 0 && p.positionFetchedAt.After(time.Now().Add(-p.positionValidFor)) {
+		p.logger.Warn("failed to fetch position from redis, falling back to last know position", "last_known", p.position)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:284]: Changed 👍

**Comment by stevesg** [comment]
One other comment - we should think about testing for this. How about a unit test for the peer? Or is there an integration test we could extend?

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 7]
```diff
@@ -0,0 +1,590 @@
...
+func (p *redisPeer) receiveLoop(name string, channel *redis.PubSub) {
+	for {
+		select {
+		case <-p.shutdownc:
+			return
+		default:
+			data, err := channel.ReceiveMessage(context.Background())
+			var opErr *net.OpError
+			if errors.As(err, &opErr) {
+				p.logger.Error("network error, waiting 10 seconds before retry", "err", err, "channel", p.withPrefix(name))
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:401]: Changed 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 8]
```diff
@@ -0,0 +1,590 @@
...
+func (p *redisPeer) fullStateReqReceiveLoop() {
+	for {
+		select {
+		case <-p.shutdownc:
+			return
+		default:
+			data, err := p.subs[fullStateChannelReq].ReceiveMessage(context.Background())
+			var opErr *net.OpError
+			if errors.As(err, &opErr) {
+				p.logger.Error("network error, waiting 10 seconds before retry", "err", err, "channel", p.withPrefix(fullStateChannelReq))
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:443]: Changed 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 9]
```diff
@@ -0,0 +1,590 @@
...
+	// lastPositionValidFor the last known position and only failover to position
+	// 0 afterwards if we do not eventually recover.
+	if len(p.Members()) == 0 && p.positionFetchedAt.After(time.Now().Add(-p.positionValidFor)) {
+		p.logger.Warn("failed to fetch position from redis, falling back to last know position", "last_known", p.position)
+		return p.position
+	}
+	for i, peer := range p.Members() {
+		if peer == p.withPrefix(p.name) {
+			p.position = i
+			p.positionFetchedAt = time.Now()
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:290]: This is not wrong, as a failed request to redis will set the members array empty. So a call to p.Members() would return an empty list, thus never running into this code.

**Comment by JohnnyQQQQ** [comment]
> One other comment - we should think about testing for this. How about a unit test for the peer? Or is there an integration test we could extend?

I went back and forth regarding testing. First I had some automated tests, but they won't add value, so I just removed them. The main problem is that you would need a redis instance to test it correctly or abstract the whole redis instance behind an interface for the sake of testing.

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 10]
```diff
@@ -0,0 +1,590 @@
...
+	readyc    chan struct{}
+	shutdownc chan struct{}
+
+	pushPullInterval time.Duration
+
+	messagesReceived     *prometheus.CounterVec
+	messagesReceivedSize *prometheus.CounterVec
+	messagesSent         *prometheus.CounterVec
+	messagesSentSize     *prometheus.CounterVec
+	nodePingDuration     *prometheus.HistogramVec
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:65]: In all honesty, I only spent a little bit of time thinking about the metrics. I copied them from what we have with Memberlist and went with that. I agree that we should also cover the other operations with metrics.

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 11]
```diff
@@ -0,0 +1,590 @@
...
+	p.messagesSent.WithLabelValues(fullState).Inc()
+	p.messagesSentSize.WithLabelValues(fullState).Add(float64(len(b)))
+	return b
+}
+
+func (p *redisPeer) Shutdown() {
+	p.logger.Info("Stopping redis peer...")
+	close(p.shutdownc)
+	p.fullStateSyncPublish()
+	del := p.redis.Del(context.Background(), p.withPrefix(p.name))
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:563]: Changed 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 12]
```diff
@@ -0,0 +1,590 @@
...
+	for {
+		select {
+		case <-p.shutdownc:
+			return
+		default:
+			data, err := channel.ReceiveMessage(context.Background())
+			var opErr *net.OpError
+			if errors.As(err, &opErr) {
+				p.logger.Error("network error, waiting 10 seconds before retry", "err", err, "channel", p.withPrefix(name))
+				time.Sleep(networkRetryInterval)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:402]: Good shout, will add it 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 13]
```diff
@@ -0,0 +1,590 @@
...
+	for {
+		select {
+		case <-p.shutdownc:
+			return
+		default:
+			data, err := p.subs[fullStateChannelReq].ReceiveMessage(context.Background())
+			var opErr *net.OpError
+			if errors.As(err, &opErr) {
+				p.logger.Error("network error, waiting 10 seconds before retry", "err", err, "channel", p.withPrefix(fullStateChannelReq))
+				time.Sleep(networkRetryInterval)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:444]: Changed 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 14]
```diff
@@ -0,0 +1,590 @@
...
+	for {
+		select {
+		case <-p.shutdownc:
+			return
+		default:
+			data, err := p.subs[fullStateChannel].ReceiveMessage(context.Background())
+			var opErr *net.OpError
+			if errors.As(err, &opErr) {
+				p.logger.Error("network error, waiting 10 seconds before retry", "err", err, "channel", p.withPrefix(fullStateChannel))
+				time.Sleep(networkRetryInterval)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:473]: Changed 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 15]
```diff
@@ -0,0 +1,590 @@
...
+	channel string
+	msgType string
+}
+
+func (c *RedisChannel) Broadcast(b []byte) {
+	b, err := proto.Marshal(&clusterpb.Part{Key: c.key, Data: b})
+	if err != nil {
+		return
+	}
+	pub := c.p.redis.Publish(context.Background(), c.channel, string(b))
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go]: It makes a lot of sense, I changed that and also added the channels part in its own file 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 16]
```diff
@@ -0,0 +1,590 @@
...
+	}
+}
+
+func (p *redisPeer) receiveLoop(name string, channel *redis.PubSub) {
+	for {
+		select {
+		case <-p.shutdownc:
+			return
+		default:
+			data, err := channel.ReceiveMessage(context.Background())
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go]: I changed the code (https://github.com/grafana/grafana/pull/65267/commits/ace72114ed6a01c805154415a08bc64608855b9f) to use the high-level API. I was using the low level API, which also makes everything hackier. 

Brief description of the high-level Channel API:

> Channel returns a Go channel for concurrently receiving messages. The channel is closed together with the PubSub. If the Go channel is blocked full for 30 seconds the message is dropped. Receive* APIs can not be used after channel is created.
> 
> go-redis periodically sends ping messages to test connection health and re-subscribes if ping can not not received for 30 seconds.

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go (hunk 4 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/multiorg_alertmanager.go:105]: Only Redis will start, as you can only have one cluster peer per Grafana instance.

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 17]
```diff
@@ -0,0 +1,590 @@
...
+	// List of active members of the cluster. Should be accessed through the Members function.
+	members    []string
+	membersMtx sync.Mutex
+
+	// Last known position in the cluster.
+	position int
+	// The time when we computed the position in the cluster the last time successfully.
+	positionFetchedAt time.Time
+	// The duration we want to return the position if the network is down.
+	positionValidFor time.Duration
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:76]: This makes sense. It also reduces the complexity. Will change it 👍

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go (hunk 2 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/multiorg_alertmanager.go]: Done 👍

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go (hunk 3 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/multiorg_alertmanager.go:103]: Done 👍

**Review by stevesg — COMMENTED** [review]
Looking good, just a handful of nits, I really like the addition of the test, it would certainly be worth trying to write a test for the peer itself (particularly the members list parts), but we could consider that in a follow up PR.

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+			p.statesMtx.RUnlock()
+
+			if !ok {
+				continue
+			}
+			if err := s.Merge(part.Data); err != nil {
+				p.logger.Warn("error merging the received broadcast message", "err", err, "key", name)
+				continue
+			}
+			p.logger.Debug("partial state was successfully merged", "key", name)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go]: done 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 19]
```diff
@@ -0,0 +1,552 @@
...
+					if !ok {
+						p.logger.Warn("received", "unknown state key", "len", len(data.Payload), "key", part.Key)
+						continue
+					}
+					if err := s.Merge(part.Data); err != nil {
+						p.logger.Warn("error merging the received remote state", "err", err, "key", part.Key)
+						return
+					}
+				}
+				p.logger.Debug("full state was successfully merged")
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go]: done 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 20]
```diff
@@ -0,0 +1,590 @@
...
+	return p.prefix + str
+}
+
+func (p *redisPeer) heartbeatLoop() {
+	ticker := time.NewTicker(p.heartbeatInterval)
+	for {
+		select {
+		case <-ticker.C:
+			startTime := time.Now()
+			cmd := p.redis.Set(context.Background(), p.withPrefix(p.name), time.Now().Unix(), time.Minute*5)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:216]: Added that 👍

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 21]
```diff
@@ -0,0 +1,568 @@
...
+
+	var part clusterpb.Part
+	if err := proto.Unmarshal([]byte(buf), &part); err != nil {
+		p.logger.Warn("error decoding the received broadcast message", "err", err)
+		return
+	}
+
+	p.statesMtx.RLock()
+	s, ok := p.states[part.Key]
+	p.statesMtx.RUnlock()
```
Comment (stevesg) [inline pkg/services/ngalert/notifier/redis_peer.go:435]: Is `cluster.State` thread safe? I think we need to hold onto the lock here.

```suggestion
	p.statesMtx.RLock()
	s, ok := p.states[part.Key]
	defer p.statesMtx.RUnlock()
```

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 22]
```diff
@@ -0,0 +1,590 @@
...
+	fullState            = "full_state"
+	fullStateChannel     = fullState
+	fullStateChannelReq  = fullStateChannel + ":request"
+	update               = "update"
+	redisServerLabel     = "redis-server"
+	networkRetryInterval = time.Second * 10
+	membersSyncInterval  = time.Second * 5
+)
+
+type redisPeer struct {
```
[the remark this answers is not in the document]
Comment (stevesg) [reply pkg/services/ngalert/notifier/redis_peer.go:46]: We could just make these `const` instead and decide whether to make them configurable in the future.

**Review by stevesg — APPROVED** [review]
Thanks for addressing all the fiddly comments, looks good from my perspective.

**Review by gotjosh — APPROVED** [review]
LGTM

**Review by alexweav — APPROVED** [review]

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go [hunk 23]
```diff
@@ -72,14 +72,44 @@ func NewMultiOrgAlertmanager(cfg *setting.Cfg, configStore AlertingStore, orgSto
...
+		redisPeer, err := newRedisPeer(redisConfig{
+			addr:     cfg.UnifiedAlerting.HARedisAddr,
+			name:     cfg.UnifiedAlerting.HARedisPeerName,
+			prefix:   cfg.UnifiedAlerting.HARedisPrefix,
+			password: cfg.UnifiedAlerting.HARedisPassword,
+			username: cfg.UnifiedAlerting.HARedisUsername,
+			db:       cfg.UnifiedAlerting.HARedisDB,
+		}, clusterLogger, moa.metrics.Registerer, cfg.UnifiedAlerting.HAPushPullInterval)
+		if err != nil {
+			return fmt.Errorf("unable to initialize redis: %w", err)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/multiorg_alertmanager.go:100]: It shuts down.

File: pkg/services/ngalert/notifier/redis_peer.go [hunk 24]
```diff
@@ -0,0 +1,564 @@
...
+		}
+	}
+	p.logger.Warn("failed to look up position, falling back to position 0")
+	return 0
+}
+
+// Returns the known size of the Cluster. This also includes dead nodes that
+// haven't timeout yet.
+func (p *redisPeer) ClusterSize() int {
+	scan := p.redis.Scan(context.Background(), 0, p.withPrefix(peerPattern), 100)
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:321]: It's the same as here https://github.com/grafana/grafana/pull/65267/files#diff-e8131ae6dd0db67fbae8188cbc95aaffe9e149c4d2d7f9a75e1edcf50cde5971R242-R245

[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/notifier/redis_peer.go:321]: And yes, the max. cluster size would be 100.

**Pull request merged**